docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,48 @@
+
+## Scenarios
+List of few scenarios where dynamic loading of full components is required:
+* MSBuild tasks - all tasks in MSBuild are dynamically loaded. Some tasks come with additional dependencies which can collide with each other or MSBuild itself as well.
+* Roslyn analyzers - similar to MSBuild tasks, the Roslyn compiler dynamically loads analyzers which are separate components with potentially conflicting dependencies.
+* XUnit loading tests - the test runner acts as an app and the test is loaded dynamically. The test can have any number of dependencies. Finding and resolving those dependencies is challenging.
+* ASP .NET's `dotnet watch` - ability to dynamically reload an app without restarting the process. Each version of the app is inherently in collision with any previous version. The old version should be unloaded.
+
+In lot of these cases the component which is to be loaded dynamically has a non-trivial amount of dependencies which are unknown to the app itself. So the loading mechanism has to be able to resolve them.  
+The SDK tooling uses the `.deps.json` format to describe dependencies of a component, so it would be beneficial if the dynamic load was able to use this same format for dependency resolution.
+
+## Dynamic loading with dependencies
+We propose to add a new public API which would dynamically load a component with these properties:
+* Component is loaded in isolation from the app (and other components) so that potential collisions are not an issue
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies.
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+* `static Assembly Assembly.LoadFileWithDependencies` - in its core similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+* `static AssemblyLoadContext AssemblyLoadContext.CreateForAssemblyWithDependencies(string assemblyPath)` - "advanced" version which would return an ALC instance and not just the main assembly. Could have overloads for enabling unloadability.
+
+## High-level description of the solution
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
```

Comment:
Did we consider a non-*.deps.json scenario where there is just a loose set of files?  dotnet/coreclr issue 13277 has the customer scenario for that case.

Reply:
Most of the dotnet/coreclr#13277 seems to actually want us to use `.deps.json` and the current lack of that functionalisty is the most blocking problem.

As for the case of no `.deps.json` and files in the same folder, which is also mentioned in that issue - that is actually described here:
> Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies. 

In short - if there's no `.deps.json` the ALC will effectively turn into "load next to me" ALC. It will behave as if there was a `.deps.json` listing all the files in the same folder and nothing else.

As for the NuGet problems in the issue - Our current solution is the `.deps.json` in combination with `.runtimeconfig.dev.json`. There is room for future improvements which would allow some of this configuration to be done programatically without the files on disk. But I personally haven't seen a user case which would need that, so I think we should postpone that for now.